File: 46_Permutations.java
import java.util.*;

class Solution {
    public List<List<Integer>> permute(int[] nums) {
        
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        
        backtrack(0, nums, result);
        
        return result;
    }
    
    private void backtrack(int pickIndex, int[] nums, List<List<Integer>> result) {
        if (pickIndex == nums.length) {
            
            List<Integer> list = new ArrayList<Integer>();
            
            for (int i = 0; i < nums.length; i++) {
                list.add(nums[i]);
            }
            
            result.add(list);
            return;
        }
        
        
        for (int i = pickIndex; i < nums.length; i++) {
            
            swap(nums, pickIndex, i);
            
            backtrack(pickIndex + 1, nums, result);
            
            swap(nums, i, pickIndex);
            
        }
        
        
    }
    
    private void swap(int[] nums, int index1, int index2) {
        if (index1 == index2) return;
        
        int temp = nums[index1];
        nums[index1] = nums[index2];
        nums[index2] = temp;
    }
}

File: 46_Permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
  
    // [1, 2, 3] // [1, 3, 2]
    
    // [2, 1, 3] // [2, 3, 1]
    
    // [3, 2, 1] // [3, 1, 2]
    const result = [];
    
    function backtrack(pickIndex) {
        if (pickIndex === nums.length) {
            result.push([...nums]); // nums.slice(0);
            return;
        }
        
        for (let i = pickIndex; i < nums.length; i++) {
            // swap
            [nums[pickIndex], nums[i]] = [nums[i], nums[pickIndex]];
            
            backtrack(pickIndex + 1);
            
            // recover
            [nums[i], nums[pickIndex]] = [nums[pickIndex], nums[i]];
        }
    }
    
    backtrack(0);
    
    return result;
};